<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-данных-и-библиотек" data-toc-modified-id="Импорт-данных-и-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт данных и библиотек</a></span></li><li><span><a href="#Первичный-анализ" data-toc-modified-id="Первичный-анализ-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Первичный анализ</a></span></li><li><span><a href="#Делим-на-целевой-и-остальные-выборки" data-toc-modified-id="Делим-на-целевой-и-остальные-выборки-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Делим на целевой и остальные выборки</a></span></li><li><span><a href="#Подвывод" data-toc-modified-id="Подвывод-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Подвывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><span><a href="#Подвывод" data-toc-modified-id="Подвывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Подвывод</a></span></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#До-преобразования" data-toc-modified-id="До-преобразования-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>До преобразования</a></span></li><li><span><a href="#После-преобразования" data-toc-modified-id="После-преобразования-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>После преобразования</a></span></li><li><span><a href="#Подвывод" data-toc-modified-id="Подвывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Подвывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Задача:
Разработка модели анонимизации персональных данных

## Загрузка данных

### Импорт данных и библиотек

In [ ]:
# Импортируем библиотеки
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression

In [ ]:
# Загружаем данные
try:
    data = pd.read_csv('/datasets/insurance.csv')
except:
    data = pd.read_csv('D:\DataHome\insurance.csv')

### Первичный анализ

In [ ]:
#Смотрим как выглядят сами данные
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [ ]:
#Изучаем данные
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


В нашем распоряжении 5 тысяч строк и 5 колонок, одна из которых булинная. Пропусков нет. Форматы данных корректны.

In [ ]:
# Смотрим основную описательную статистику
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

В наших данные поровну мужчин и женщин, средний и медианный возраст равняется 30, значит данные не скошены. Также нет скошенности и в зарплате, средняя и медианна зп клиентов равняется примерно 40 тысячам. В среднем у семей по одному ребенку, но где-то есть и все 6. Большинство клиентов не получали страховых выплат, но у кого-то были и целых 5 выплат

In [ ]:
data['Страховые выплаты'].value_counts()

0    4436
1     423
2     115
3      18
4       7
5       1
Name: Страховые выплаты, dtype: int64

Как мы уже заметили, большинство клиентов не получали страховых выплат, одну страховую выплату получили 423 клиента, 2 выплаты 115 клиентов и так далее. Такое распределение выглядит логичным, с ним можно работать

In [ ]:
#Смотрим корреляции
data.corr()

Пол   Возраст  Зарплата  Члены семьи  \
Пол                1.000000  0.002074  0.014910    -0.008991   
Возраст            0.002074  1.000000 -0.019093    -0.006692   
Зарплата           0.014910 -0.019093  1.000000    -0.030296   
Члены семьи       -0.008991 -0.006692 -0.030296     1.000000   
Страховые выплаты  0.010140  0.651030 -0.014963    -0.036290   

                   Страховые выплаты  
Пол                         0.010140  
Возраст                     0.651030  
Зарплата                   -0.014963  
Члены семьи                -0.036290  
Страховые выплаты           1.000000

Слишком высоких корреляций нет

### Делим на целевой и остальные выборки

In [ ]:
#Делим на основные и целевой признаки
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']
features.shape, target.shape

((5000, 4), (5000,))

In [ ]:
#Делим на обучающие и валидационные выборки
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=12345
    )
features_train.shape, features_valid.shape, target_train.shape, target_valid.shape

((3750, 4), (1250, 4), (3750,), (1250,))

Данные разделились как надо

### Подвывод
Мы провели первичный анализ данных, вывели, что пропусков нет, формат данных корректен, как и сами данные, скошенности не наблюдалось ни в одном признаке, проблемы мультиколлинеарности нет.

## Алгоритм преобразования

**Алгоритм:** Всё что нам нужно сделать для шифрования, это сместить плоскость матрицы на случайные, известные только нам, координаты. Это будет работать на свойстве обратных матриц:
$$
MM^{-1} = M^{-1}M = E
$$

$E$ - это единичная матрица, можно назвать это статусом-кво для базисных векторов матрицы: $x = \begin{bmatrix} 1 \\ 0 \\ 0 \end{bmatrix} , y= \begin{bmatrix} 0 \\ 1 \\ 0 \end{bmatrix}, z = \begin{bmatrix} 0 \\ 0 \\ 1 \end{bmatrix}$


Для шифровки мы умножаем матрицу признаков на случайную матрицу, которая будет ключом. При умножении на обратную ключу матрицу, в соответствии показанному выше свойству, матрица вернётся к исходному положение, как и наши данные

In [ ]:
#Наш "ключ" для шифрования
np.random.seed(seed=12345)

In [ ]:
#Создаём случайную матрицу 5 на 5
rand_m = np.random.normal(0, 5, (5, 5))
#Создаём обратную ей матрицу
rand_m_reversed = np.linalg.inv(rand_m)
rand_m, rand_m_reversed, np.dot(rand_m, rand_m_reversed)

(array([[ -1.0235383 ,   2.39471669,  -2.59719358,  -2.77865152,
           9.82890286],
        [  6.96702916,   0.46453938,   1.40873076,   3.84511284,
           6.23217368],
        [  5.03594679,  -6.48110555,   1.37495817,   1.14456439,
           6.76458418],
        [  4.4321467 , -10.00818655,  -1.85921269,   8.34512655,
          -2.19284868],
        [ -2.69870723,   2.38492505,  16.2447196 ,  -5.10613762,
          -2.88543652]]),
 array([[-0.22408388,  0.07201981,  0.16884577, -0.17789924, -0.07672365],
        [ 0.00609462,  0.0927049 , -0.10372411, -0.02732391, -0.00141289],
        [ 0.03967309,  0.02724685, -0.03457678,  0.05085067,  0.074285  ],
        [ 0.17081662,  0.08310047, -0.22999878,  0.20879633,  0.06346758],
        [ 0.13569421,  0.01560557, -0.0313035 ,  0.06059553,  0.0299261 ]]),
 array([[ 1.00000000e+00,  1.65375912e-17,  4.37390849e-17,
          8.82270549e-17,  2.05656504e-18],
        [ 5.80138943e-17,  1.00000000e+00,  2.96404754e-16,
         -4.

In [ ]:
#Как выглядит данные после шифровки
data_dot_rand_m = data @ rand_m
data_dot_rand_m.columns = data.columns
data_dot_rand_m

Пол        Возраст      Зарплата   Члены семьи  \
0     250072.017485 -321451.402415  68251.226646  56933.610077   
1     191688.194722 -246268.265181  52327.597474  43673.561177   
2     105956.926392 -136089.744815  28914.974703  24147.360561   
3     210154.152962 -270272.362296  57361.620491  47825.772881   
4     131632.264443 -169141.452920  35923.255431  29978.015209   
...             ...            ...           ...           ...   
4995  179987.241439 -231382.477247  49121.732604  40985.302303   
4996  264124.922825 -339604.144435  72093.845593  60114.253244   
4997  170866.800987 -219720.203580  46635.538057  38894.325489   
4998  164841.007508 -211949.561316  44983.949311  37534.104916   
4999  204657.925014 -263127.491517  55858.289642  46582.544058   

      Страховые выплаты  
0         335786.530296  
1         257335.800386  
2         142237.000729  
3         282209.650082  
4         176739.976755  
...                 ...  
4995      241665.770243  
4996      354673.911870  
4997      229439.661337  
4998      221342.260727  
4999      274824.254456  

[5000 rows x 5 columns]

In [ ]:
#Как выглядят расшифрованные данные
data_dot_rand_m_reversed = data_dot_rand_m @ rand_m_reversed
data_dot_rand_m_reversed.columns = data.columns
data_dot_rand_m_reversed

Пол  Возраст  Зарплата   Члены семьи  Страховые выплаты
0     1.000000e+00     41.0   49600.0  1.000000e+00      -4.256383e-13
1     8.113634e-12     46.0   38000.0  1.000000e+00       1.000000e+00
2     5.204800e-12     29.0   21000.0  2.137293e-13       6.454238e-13
3     6.815135e-12     21.0   41700.0  2.000000e+00       8.408270e-13
4     1.000000e+00     28.0   26100.0 -8.558991e-13       3.233051e-13
...            ...      ...       ...           ...                ...
4995  1.079101e-11     28.0   35700.0  2.000000e+00       3.542638e-12
4996  5.253163e-12     34.0   52400.0  1.000000e+00      -3.384161e-13
4997  5.796234e-12     20.0   33900.0  2.000000e+00       1.476578e-13
4998  1.000000e+00     22.0   32700.0  3.000000e+00      -4.472130e-13
4999  1.000000e+00     28.0   40600.0  1.000000e+00      -1.036553e-12

[5000 rows x 5 columns]

In [ ]:
#Делим на основные и целевой признаки
features_rand = data_dot_rand_m_reversed.drop(['Страховые выплаты'], axis=1)
target_rand = data_dot_rand_m_reversed['Страховые выплаты']
features_rand.shape, target_rand.shape

((5000, 4), (5000,))

In [ ]:
#Делим на обучающие и валидационные выборки
features_train_rand, features_valid_rand, target_train_rand, target_valid_rand = train_test_split(
    features_rand, target_rand, test_size=0.25, random_state=12345
    )
features_train_rand.shape, features_valid_rand.shape, target_train_rand.shape, target_valid_rand.shape

((3750, 4), (1250, 4), (3750,), (1250,))

### Подвывод
Мы описали алгоритм шифрования, сгенерировали псевдослучайную матрицу, а также матрицу ей обратную, разделили на выборки преобразованные данные для последующего использования в линейной регрессии.

## Проверка алгоритма

### До преобразования

In [ ]:
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
r2s = r2_score(target_valid, predictions)
r2s

0.43522757127026546

Результат нашей метрики 0.43522. В результате шифрования, модель должна выдавать те же результаты, но данные сами должны быть преобразованными.

### После преобразования

In [ ]:
model = LinearRegression()
model.fit(features_train_rand, target_train_rand)
predictions = model.predict(features_valid_rand)
r2s = r2_score(target_valid_rand, predictions)
r2s

0.4352275712700978

Результат тот же, что и до преобразования

### Подвывод
Мы придумали алгоритм шифрования данных и проверили, что шифровка не влияет на точность работы линейной регрессии

## Вывод

Мы провели предобработку данных, провели первичный анализ, придумали алгоритм шифрования данных и сравнили качество линейной регрессии на изначальных и преобразованных матрицах. Как и ожидалось, качество матрицы осталось тем же, значит поставленную задачу мы успешно выполнили.